connections/simopen.md

```diff
@@ -55,18 +59,30 @@ A ---> B: iamclient
 B ---> A: iamclient
 A: generate random integer IA
 B: generate random integer IB
-A ---> B: {IA}
-B ---> A: {IB}
+A ---> B: select:{IA}
+B ---> A: select:{IB}
 A ---> B: initiator
 B ---> A: responder
 ```
 
 In the unlikely case where both peers selected the same integer, they
-generate a fresh one and enter another round of the protocol.
+generate a fresh one and enter another round of the protocol. If
```

Comment:
In the case where both peers select the same integer could they use their peer ids as a tie-breaker (eg lowest wins)?

Reply:
multistream has no concept of peer ids which prevents us from doing this, otherwise we could avoid the random number altogether.

Reply:
This look a bit like an implementation detail to me.
Rust-libp2p is using the lowest `PeerId` to determine who has priority in case of simultaneous connection.

Reply:
We can make this an arbitrary byte string. The advantage of using a peer ID here is that, later (after handshaking), we _could_ validate if the peer was honest, and disconnect/blacklist it if dishonest. Rationale: if I'm dishonest, I can make myself win this challenge always by sending 0. Not that this will gain me any advantage that I can think of, but it's a nuance worth considering.

Reply:
We could also mitigate this by making peers hash the received nonce (e.g. SHA256) prior to comparison.

Reply:
@tomaka curious to learn you already have an implementation. Mind posting a pointer to the code?

Reply:
I don't think we should even have the concept of a peer id in multistream, this is at the bottom of the stack.
We could use hashing, but does it really gain anything?

Reply:
> @tomaka curious to learn you already have an implementation. Mind posting a pointer to the code?

Rust-libp2p doesn't implement UDP or hole punching, so we're not doing this in the scope of what this PR is about. Rust-libp2p only ever allows one connection between two given peers, and if we detect two active connections, the one with the lowest `PeerId` has the dialing priority and the other one is shut down.

Here's the code that handles an incoming connection for which we learned the `PeerId` of the remote: https://github.com/libp2p/rust-libp2p/blob/2fd941122a6c589501ad1c04b0f091f1337f5f49/core/src/nodes/network.rs#L1130-L1196

Reply:
This discussion seems to have stalled. The spec has settled on a random integer instead of a peer ID. Please un-resolve in case you would like to continue this discussion.